File: python/problem_82.py
class Reader82:
    def __init__(self, content: str):
        self._content = content
        self._pos = 0
        self._buf = ''

    def _read7(self) -> str:
        chunk = self._content[self._pos:self._pos + 7]
        self._pos += len(chunk)
        return chunk

    def read_n(self, n: int) -> str:
        while len(self._buf) < n:
            chunk = self._read7()
            if not chunk:
                break
            self._buf += chunk
        result = self._buf[:n]
        self._buf = self._buf[n:]
        return result


def main():
    print("Problem 82:")
    r = Reader82("Hello world")
    assert r.read_n(5) == "Hello"
    assert r.read_n(1) == " "
    assert r.read_n(5) == "world"
    assert r.read_n(5) == ""
    print("All Tests Passed")

File: lua/problem_82.lua
local M = {}

function M.new_reader(content)
    local r = {content=content, pos=1, buf=""}
    function r:read7()
        local s = self.content:sub(self.pos, self.pos+6)
        self.pos = self.pos + #s
        return s
    end
    function r:read_n(n)
        while #self.buf < n do
            local chunk = self:read7()
            if #chunk == 0 then break end
            self.buf = self.buf .. chunk
        end
        local result = self.buf:sub(1, n)
        self.buf = self.buf:sub(n+1)
        return result
    end
    return r
end

function M.problem82()
    print("Problem 82:")
    local r = M.new_reader("Hello world")
    assert(r:read_n(5) == "Hello")
    assert(r:read_n(1) == " ")
    assert(r:read_n(5) == "world")
    assert(r:read_n(5) == "")
    print("All Tests Passed")
end

return M
